### PR TITLE
Add ci

### LINKED ISSUE #54: add CI option
see #49 - this new option was provided by a volunteer and needs to be discussed / confirmed with core admin team

### PR DESCRIPTION
**Pull request by jhoover4**
Adding CI functionality, prettier, and eslint. For this to work we have to move everything to root so the project is not `/src`.

## Prettier

Prettier is a code formatter used by the javascript community to automatically enforce one code style so there is no arguing about "spaces/tabs" or single quotes or double quotes, etc. It leaves what we call "bike shedding" (arguing about dumb pointless things) at the door. See https://prettier.io/ for more details.

## Eslint

Eslint is used to check code for common errors and automatically correct them. This keeps it easy for reviewers to focus on more complex issues and helps you quickly fix issues. See https://eslint.org/ for more details. We're using the React extension here, which is a common extension for react apps.

## Continuous Integration

Continous Integration is what developers call the automatic process of linting (checking code for style errors or erros in general), testing, and other checks BEFORE your branch is merged into master (integrated). This is done usually through platforms that are running "jobs" for these various tasks. See https://docs.gitlab.com/ee/ci/introduction/ for a good overview. Here we're using https://docs.github.com/en/actions/using-workflows as our CI runner to execute these tasks for us.

### CODE REVIEW

File: src/index.js
Comment (ProsperousHeart) [inline src/index.js:1]: updated for linting

File: src/reportWebVitals.js
Comment (ProsperousHeart) [inline src/reportWebVitals.js:1]: updated for linting - approved

File: src/setupTests.js
Comment (ProsperousHeart) [inline src/setupTests.js:1]: updated for linting - approved

File: src/index.css
Comment (ProsperousHeart) [inline src/index.css:1]: this was moved out oy my-app into the main app folder as per [here](https://github.com/codefordallas/codefordallas.github.io/pull/49/files#diff-81a3a4ae01a845f636b80385ed1762d8b260da996657d338f15c31ec15ff075c)

File: my-app/src/components/Projects/Projects.css
Comment (ProsperousHeart) [inline my-app/src/components/Projects/Projects.css:1]: moved to from **my-app** to main app folder (parent) for #50 - approved

File: src/components/Projects/Projects.css
Comment (ProsperousHeart) [inline src/components/Projects/Projects.css:1]: moved to from **my-app** to main app folder (parent) for #50 - approved

File: package.json
Comment (ProsperousHeart) [inline package.json]: why was this removed?

File: .prettierignore
Comment (jhoover4) [inline .prettierignore]: This file is telling prettier what NOT to lint for styling. See [https://prettier.io/docs/en/ignore.html](https://prettier.io/docs/en/ignore.html)

File: .prettierrc.json
Comment (jhoover4) [inline .prettierrc.json]: This file is for any additional prettier configuration. See [https://prettier.io/docs/en/configuration.html](https://prettier.io/docs/en/configuration.html)

File: package.json
Comment (jhoover4) [inline package.json]: There are 4 scripts being added here. 2 make updates using eslint and prettier and 2 just perform checks and return either a 0 or 1 (success or failure)

File: package.json
Comment (jhoover4) [inline package.json]: Here we're adding dependencies that dont need to be bundled with the production app and are only used for developers when they're making code. See [https://docs.npmjs.com/specifying-dependencies-and-devdependencies-in-a-package-json-file](https://docs.npmjs.com/specifying-dependencies-and-devdependencies-in-a-package-json-file)

File: .github/workflows/ci.yaml
Comment (jhoover4) [inline .github/workflows/ci.yaml:1]: See my PR overview for what CI is. This file is what github uses to define its "actions" (the word gitlab uses for runners). See [https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions)

File: .github/workflows/ci.yaml
Comment (jhoover4) [inline .github/workflows/ci.yaml:3]: When will all this run? On pushes to master (if someone pushed directly) and on PRs to the `v1-mvp` branch

File: .github/workflows/ci.yaml
Comment (jhoover4) [inline .github/workflows/ci.yaml:9]: What individual jobs will be running every time? Each job is a separate task to be done.

File: .github/workflows/ci.yaml
Comment (jhoover4) [inline .github/workflows/ci.yaml:11]: The underlying OS we'll be using is the latest version of `ubuntu`, which is an open source linux operating system. See [https://ubuntu.com/](https://ubuntu.com/)

File: .github/workflows/ci.yaml
Comment (jhoover4) [inline .github/workflows/ci.yaml:15]: We want to use the latest version of node, but we dont care about minor version differences

File: .github/workflows/ci.yaml
Comment (jhoover4) [inline .github/workflows/ci.yaml]: I can delete this, but its commented out for when I add testing later

File: .github/workflows/ci.yaml
Comment (jhoover4) [inline .github/workflows/ci.yaml:18]: Get the repo code first

File: .github/workflows/ci.yaml
Comment (jhoover4) [inline .github/workflows/ci.yaml:24]: What will we call this job?

File: .github/workflows/ci.yaml
Comment (jhoover4) [inline .github/workflows/ci.yaml:26]: This runs the npm script I made, then if there is an error it runs the echo command. `||` in bash only runs if there is a non-0 exit code (an error) in the previous line

File: .gitignore
Comment (jhoover4) [inline .gitignore]: We also want to ignore IDE files from git. I use Jetbrains which places its own `.idea` directory in the folder.

File: .eslintrc.js
Comment (jhoover4) [inline .eslintrc.js]: This is the eslint configuration file. See [https://eslint.org/docs/latest/use/configure/configuration-files](https://eslint.org/docs/latest/use/configure/configuration-files)

File: .eslintrc.js
Comment (jhoover4) [inline .eslintrc.js]: We want to use eslint's reccomended checks in addition to what react recommends, and prettier

File: .eslintrc.js
Comment (jhoover4) [inline .eslintrc.js]: This is for a new app `es2021` and is going to run in the browser

File: .eslintrc.js
Comment (jhoover4) [inline .eslintrc.js]: How are we importing our js code? We're using [module syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) as its the new default and standard

**Comment by ProsperousHeart** [comment]
@jhoover4 - the files have been moved as of #53. With **public** folder being removed from `.gitignore` file, that folder was also uploaded to the repo. So the error you were experiencing should no longer occur. (In the future, please let us know ASAP if you run into issues like that so we can avoid confusion, misunderstandings, issues for other volunteers, etc.)

It looks like this change introduced some conflicts to this PR - likely because in this PR the files appear to have been cut and pasted instead of moved. (See [files changed](https://github.com/codefordallas/codefordallas.github.io/pull/49/files).)

If it is easier, feel free to close this PR and create 2 new ones for:
- linting
- CI

Otherwise, you can keep this one open for CI but please pull from `v1-mvp` branch, remove the linting, and create a new PR for linting. As discussed in Discord, we work on one issue or enhancement per PR request.

Thank you!

File: .github/workflows/ci.yaml
Comment (ProsperousHeart) [inline .github/workflows/ci.yaml:15]:
```suggestion
        node-version: [18.x, 20.x]
```
I've been on v20 for a while. Any reason both are not here? (I also messaged you in Discord)

File: .github/workflows/ci.yaml
Comment (ProsperousHeart) [inline .github/workflows/ci.yaml:19]:
```suggestion
        uses: actions/checkout@v3
```
Latest version is 3 as per [here](https://github.com/marketplace/actions/checkout)

File: .github/workflows/ci.yaml
Comment (ProsperousHeart) [reply .github/workflows/ci.yaml:26]: Script you made? When going to npm site [format-check](https://www.npmjs.com/package/format-check) seems to be in another language. Is it not published? Or is this the same one?

**Review by ProsperousHeart — CHANGES_REQUESTED** [review]
Please see notes / discussions / responses, including but not limited to (since apparently I can't link to some of my own comments/reviews):
- [here](https://github.com/codefordallas/codefordallas.github.io/pull/49#issuecomment-1585184280)
- [here](https://github.com/codefordallas/codefordallas.github.io/pull/49#discussion_r1221664110)

Also curious why you used the 2 specific npm packages listed instead of:
- [eslint](https://www.npmjs.com/package/eslint) (over 32M weekly downloads and last updated June 2, 2023)
- [prettier](https://www.npmjs.com/package/prettier) (over 29M weekly downloads with last update 2 months ago)

File: .github/workflows/ci.yaml
Comment (jhoover4) [reply .github/workflows/ci.yaml:26]: This isn't another package its just an npm script. These are very common. See https://docs.npmjs.com/cli/v9/using-npm/scripts#npm-run-user-defined

File: package.json
Comment (jhoover4) [reply package.json]: Its now using the eslint config file itself instead of pulling from here

File: .github/workflows/ci.yaml
Comment (jhoover4) [reply .github/workflows/ci.yaml:15]: 18 is the latest stable, 20 is the latest nightly. See here: https://nodejs.org/en

We can use 20.x if you want.

File: .github/workflows/ci.yaml
Comment (jhoover4) [reply .github/workflows/ci.yaml:19]: Cool good catch!

**Comment by jhoover4** [comment]
@ProsperousHeart I did let you know ASAP, I'm the one who found it and fixed it. I'm confused why PRs aren't sufficient for communication here.

The merge conflict is just there because we pulled in the other PR. All you need to do is merge locally and push up, which I did.

I'm not going to make a whole other PR for linting/CI. This is an operation that is being done through the CI and makes sense as one PR. Why are we introducing more work and overhead here when the work is already done? This is already a very small PR.

File: src/App.js
Comment (ProsperousHeart) [inline src/App.js:1]: While I don't agree with some of the things that were changed, it doesn't impact the code. (And likely was from the linter.) Confirming changes.

File: .eslintrc.js
Comment (ProsperousHeart) [reply .eslintrc.js]: Adding info to documentation as per #58

File: .eslintrc.js
Comment (ProsperousHeart) [reply .eslintrc.js]: Added to documentation for #58

File: .eslintrc.js
Comment (ProsperousHeart) [reply .eslintrc.js]: Added to documentation for #58

File: .eslintrc.js
Comment (ProsperousHeart) [reply .eslintrc.js]: Added to documentation for #58

File: .prettierignore
Comment (ProsperousHeart) [reply .prettierignore]: Added to documentation for #58

File: .prettierrc.json
Comment (ProsperousHeart) [reply .prettierrc.json]: Added to documentation for #58

File: package.json
Comment (ProsperousHeart) [reply package.json]: Added to documentation for #58

**Comment by ProsperousHeart** [comment]
with #57 now complete, this one should be good to go for reintegration and final check before merging

thank you again for working with us through this

File: src/components/Projects/chatGPT-Projects.html
Comment (jhoover4) [inline src/components/Projects/chatGPT-Projects.html]: Not sure why prettier decided to run on this now but not before but here we are

File: package-lock.json
Comment (ProsperousHeart) [inline package-lock.json]: In line 176 it says we need this, but it's being deleted in the update?

File: package-lock.json
Comment (ProsperousHeart) [inline package-lock.json]: What is this???

File: src/components/Projects/chatGPT-Projects.html
Comment (ProsperousHeart) [reply src/components/Projects/chatGPT-Projects.html]: So ... whatever it changed to is partially what it was before. So that's super weird. It looks like a strange mix of the two now. I wonder if all the updates to the packages also changed what it caught?

File: src/components/Projects/chatGPT-Projects.html
Comment (ProsperousHeart) [inline src/components/Projects/chatGPT-Projects.html]: This is made wrong again, so what changed? The old stuff in red was merged from your most recent PR.

File: src/components/Projects/chatGPT-Projects.html
Comment (ProsperousHeart) [inline src/components/Projects/chatGPT-Projects.html]: this was also removed in your most recent PR ...

**Review by ProsperousHeart — CHANGES_REQUESTED** [review]
For the most part the updates were ok (though one of the packages was updated oddly) ... And while I'm weary about the package file update, I have no knowledge to voice any changes. So I would have approved, but then the changes in other files are forcing a halt.

The files that appear to have been changed from linting? This is putting the file back tot he way it was before your last linting update and merge. Your new update is trying to undo that and introduce things like a spelling error that was already rectified.

So cannot approve yet. 😞

**Comment by ProsperousHeart** [comment]
I also saw the package-lock file being the only package file updated. Any idea to that @jhoover4 ?

**Comment by jhoover4** [comment]
> I also saw the package-lock file being the only package file updated. Any idea to that @jhoover4 ?

Probably because I didnt commit it before and things have updated 🤷. It should be fine after this

File: src/components/Projects/chatGPT-Projects.html
Comment (jhoover4) [reply src/components/Projects/chatGPT-Projects.html]: Ah oops I commited by accident thanks

File: src/components/Projects/chatGPT-Projects.html
Comment (jhoover4) [reply src/components/Projects/chatGPT-Projects.html]: I just reverted

**Review by ProsperousHeart — APPROVED** [review]
Approved! Thanks a bunch :D

**Pull request merged**
